src/main/java/org/truffleruby/core/kernel/KernelNodes.java

```diff
@@ -749,8 +719,12 @@ protected Object evalBindingAddsVarsCached(
                 @Cached("createCallTarget(rootNodeToEval)") RootCallTarget cachedCallTarget,
                 @Cached("create(cachedCallTarget)") DirectCallNode callNode,
                 @Cached RopeNodes.EqualNode equalNode) {
+            final MaterializedFrame originalFrame = binding.getFrame();
             final MaterializedFrame parentFrame = BindingNodes.newFrame(binding, newBindingDescriptor);
```

Comment:
This would chain 1 frame from each `eval()` call with this Binding instance, which could be problematic.
I guess that's the reason for the `assignsNewUserVariables()` check below.
@aardvark179 Do you remember the details?

Reply:
I'm thinking this node shouldn't create any new frame, but rather just (In)DirectCallNode#call will create the frame, and we should somehow return that + the return value of the eval code.
If that returned frame has a non-empty descriptor, then we should assign the frame to the binding, if not, we should copy the state like DeclarationContext back to the Binding's frame.

Reply:
> If that returned frame has a non-empty descriptor, then we should assign the frame to the binding, if not, we should copy the state like DeclarationContext back to the Binding's frame.

What the difference of creating the new frame on the `EvalNode` level or in `(In)DirectCallNode#call ` ? We will have the same `parentFrame`, but on the level below, isn't it?

Reply:
`(In)DirectCallNode#call` always creates a new frame (Truffle does).
So I think we should not need an extra frame above.
That was (I think) just a convenient way to have local variables assigned in the Binding, but it need parsing twice to adjust the depth to search variables in the binding, which seems messy.

Reply:
So,
1. We have `public final class DefaultDirectCallNode extends DirectCallNode` which has `#call` method 
```java
    public Object call(Object... arguments) {
        return ((DefaultCallTarget) callTarget).callDirectOrIndirect(this, arguments);
    }
```

2. In `DefaultCallTarget#callDirectOrIndirect` we have:

```java
Object callDirectOrIndirect(final Node callNode, Object... args) {
   ...
  final DefaultVirtualFrame frame = new DefaultVirtualFrame(rootNode.getFrameDescriptor(), args);
  ...
  return rootNode.execute(frame);
}
```

3. Instead of creating the new frame in `EvalNode`, we need to return if from `DefaultCallTarget`, like:
```java
Object callDirectOrIndirect(final Node callNode, Object... args) {
   ...
  final DefaultVirtualFrame frame = new DefaultVirtualFrame(rootNode.getFrameDescriptor(), args);
  ...
  return [frame, rootNode.execute(frame)]; // <--- here
}
```

4. Since we cannot "patch" `DefaultCallTarget`, we need to create the new class `class NewClassThatReturnsFrameAndResult < RootCallTarget` and use (instantiate) in `EvalNode#createCallTarget`.

5. Hande returned result (aka `[frame, rootNode.execute(frame)];`) on `EvalNode` as you described above (`If that returned frame has a non-empty descriptor, then we should assign the frame to the binding, if not, we should copy the state like DeclarationContext back to the Binding's frame.`)

Am I right? 🤔

Reply:
Yeah, we should not modify Truffle code.
I'm rather thinking we could pass some boolean flag to the TranslatorDriver, and that would add an extra node wrapping the body node (somewhere around [here](https://github.com/oracle/truffleruby/blob/b6b312a345382e336229e2082a307656c2fe59be/src/main/java/org/truffleruby/parser/TranslatorDriver.java#L289-L362)) that does:
```java
RubyNode child;

Object execute(VirtualFrame frame) {
  Object result = child.execute();
  return new Object[] { result, frame.materialize() };
}
```